Add support for a structured BOM format (#166)
buildpacks/rfcs: text/0000-sbom.md

> - Reducing unplanned and unproductive work
> - Supporting more informed market differentiation and component selection
> - Reducing duplication of effort by standardizing formats across multiple sectors
> - Identifying suspicious or counterfeit software components

Currently since container images built by Cloud Native Buildpacks don't follow a generic file system layout, they may not be easily scannable by container scanning tools. We do however provide a BOM which should greatly help sidestep the entire manual scanning process and speed up things like CVE detection by directly providing the CVE scanner with a BOM.

However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.

Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.

## What use cases does it support?

- Consistent SBOM across buildpacks
- Improved integration with other supply chain analysis and scanning tools

## What is the expected outcome?

> - Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.

Reviewer: If we use CycloneDX as the internal storage format, can it be converted to the other formats in a lossless manner? My concern would be a user needing format XYZ, but that format is missing info due to the conversion process.

Have you done any conversion demos/experiments? Like here's the BOM info for a container in CycloneDX format, I can convert this to SPDX and not lose data? Or possibly, can you convert CycloneDX -> SPDX -> CycloneDX and have the original document (although we may not care about that second conversion bac)? I think including a section on conversion details might be helpful in justifying this choice.

Reviewer 2: One thing I would like to point out is that vulnerabilities are considered "dynamic" data. That is, if scanners find a vulnerability, it may not actually be something that manifests because reasons. My suggestion would be to create an internal key-value store containing a union of CycloneDX and SPDX metadata so the formats can be generated based on the use.

Reviewer 3: @dmikusa-pivotal whether or not you can losslessly convert from the CycloneDX format to others depends on what information you have. I wrote a converter for turning a CycloneDX BOM into an SPDX one and had to omit a lot of things simply because SPDX just doesn't support the same things as CycloneDX.

It's probably helpful to understand that the two formats originally came from two different needs. SPDX came from the need to address OSS license compliance. And CycloneDX came about to address security use cases because SPDX didn't.

The CycloneDX spec always supported custom extensions and namespaces. But we've added custom properties in v1.3. So you could potentially use that for storing additional information that either can't already be represented, or that is SPDX/other format specific.

Reviewer: Thanks for sharing that. I wasn't aware of the two different target use cases. I can absolutely see users wanting information for security purposes but at the same time, I can see them wanting this information for OSS compliance reasons as well. Whatever we end up doing, probably needs to support both use cases. It sounds like the CycloneDX format, used internally could accommodate that, if we make use of custom extensions/namespaces/properties.

I think it would still be interesting to see some examples of how this might work out for an image. Kind of a PoC to prove it out.

Reviewer 3: Don't get me wrong. There is some overlap in the formats. My comment was more about where the formats have come from. And to explain that they both have strengths and weaknesses because of that.

CycloneDX already supports OSS license use cases that can be automated. But SPDX also supports some manual audit scenarios like https://spdx.github.io/spdx-spec/3-package-information/#316-comments-on-license and https://spdx.github.io/spdx-spec/4-file-information/#45-concluded-license

And from what I've heard SPDX is starting to work on support for more security use cases. Like last year they added support for package URL as an option for identifying a package.

Reviewer 4: +1 to @dmikusa-pivotal's point about seeing a demo of some sort of the format conversion process. 

From a buildpack author's perspective, I know the set of metadata I have the ability to pass along to the BoM, but from looking at both formats it's not abundantly clear what the best way to represent that metadata is to minimize data loss.  The benefit of having the data in the format now is that we have the full set of metadata, and other tools can take that raw metadata into the best structure for the requested format. I worry about the data loss that comes along with supporting multiple formats.

Author: I have included some of the conversions below for reference. Apart from that, this RFC draft currently delegates the responsibility for conversion between SBOM formats to platforms so we could have the tooling that converts these fields losslessly either in `pack` or cli tools for other platforms like `kp cli`.

Author: Also to note  - CDX currently has a free for key value store (albeit the values have to be string only)
This can be added to any component in the SBOM - https://cyclonedx.org/use-cases/#properties--name-value-store
```json
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.3",
  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
  "version": 1,
  "metadata": {
    "properties": [
      {
        "name": "Foo",
        "value": "Bar"
      }
    ]
  },
  "components": [
    {
      "type": "library",
      "name": "acme-library",
      "version": "1.0.0",
      "properties": [
        {
          "key": "Foo",
          "value": "456"
        }
      ]
    }
  ]
}
```

This can be used to store all this metadata (this would essentially be the same as storing it in the freeform metadata table in the current BOM spec). Whatever tooling was originally parsing out the free-form BOM table could be accommodated to look at these fields while at the same time we would get benefits from auto-generating the SBOM for components not installed directly through tarballs.

Author: From a quick scan at https://spdx.github.io/spdx-spec/  I couldn't find an equivalent in SPDX. Although I may not be looking in the right place.

Reviewer 4: RE: defining the fields with someone from the Paketo team, @ForestEckhardt and I can do that with you for sure.

- Moving the SBOM from a label to a file in a separate layer.
- Enhancements to the platform spec and `pack` to allow easy interaction with SBOMs including conversion between major SBOM formats.

# What it is
[what-it-is]: #what-it-is

Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.

> It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` may be `xml` or `json`. This is mostly to allow for different BOM generation tools to be used. This also allows us in the future to add support for `spdx`  tag documents or other file formats if needed.

Reviewer: >where `<ext>` may be `xml` or `json`

It seems that having support for multiple data representations would put extra burden on the lifecycle to understand all of those, convert, merge, and ultimately pull them into one representation in the CycloneDX format.

What about picking one standard format like "json", and pushing the formatting/conversion responsibilities to the integrator/buildpack (`libcnb` can probably help some with minimizing the work on an author too).

For example, if I want to use tool XYZ that only outputs in XML format. I as the buildpack author would be responsible for running the tool to get the SBOM and then converting it to JSON. The benefit of this, IMHO, would be that tool format/conversion issues would be on the user of the tool to fix, not on the lifecycle team.

Reviewer 2: 100% this

Reviewer 3: FYI CycloneDX and SPDX tooling already exists to convert between JSON and XML representations.

Author: This makes sense to me. I guess JSON might be an easier target give that our current SBOM label is also json. I guess we can adopt a naming scheme as suggested in the NTIA document around 

`sbom.spdx.json` or `sbom.cdx.json` (I believe the cdx one was a suggesting but not a standard?)

Reviewer 2: Tern uses an internal data model to represent all the keys and then uses plugins to generate each of the formats based on a internal-model-key to format-key template. Hence there is no data loss when generating each format. The loss happens when reading the formats for obvious reasons.

The above bom documents need to be created in CycloneDX format.

The lifecycle of the respective `bom` files during rebuilds would be exactly the same as the `bom` table currently. 

> The lifecycle would be responsible for taking all of the above `bom` files, merging it and storing it in `/layers/config/sbom.json`. This file would also be stored in a layer of its own and the lifecycle would be responsible for storing the layer digest in a label `io.buildpacks.bom.digest`.

Reviewer: Merging the files makes me worry a bit (maybe unnecessarily?). I start thinking about potential edge cases. Maybe it's unnecessary. What if each layer just had its own individual file. Then no merging is required. Instead of a single file like `sbom.json`, we could have `/layers/config/sbom/<layer>.json`.

I guess the downside of this would be that scanners would need to process multiple files, but that seems like something reasonable. It also seems like it would provide more detail, because the scanner could then know specifically which layer (and could trace back to the buildpack) that installed a particular thing. Maybe they could figure that out another way, but this would make it obvious.

Reviewer 3: I don't know enough about this tooling. So I might be missing the mark here.

Layer level BOMs could be beneficial for automated analysis of each layer.

You can also create a nested BOM. With multiple layers as components under the top level container image.

And yet another option is to create separate BOMs for each layer. And then one final BOM that references all the individual layer BOMs.

They all have different pros and cons. But as someone using a particular container I probably don't care as much about the individual layers. I want to know what I'm using. And I most likely want it as a single BOM.

As someone building a container I probably want to know more per layer detail.

So a combination of the two could be beneficial depending on the use cases you are trying to serve.

Reviewer: That's interesting and very flexible. I think my primary concern would be for the support of various tools that will consume this. I see it less as humans reading it and much more as automated tools procured by humans reading it.

Are all of these different ways to represent the layers in the  BOM required to be supported by the CycloneDX specification? Or are some optional under the spec? My worry would be picking a way to represent it, then having popular tool XYZ not support that part of the spec so be unable to read the BOM correctly. Thanks.

Reviewer 3: Yeah, most of the time it should be automated tools.

We don't put restrictions on implementations. It's an open standard that anyone can contribute to, adopt and use. I don't think it's reasonable for us to say you must do X, Y and Z if you are implementing CycloneDX.

But in terms of the simplest, most well supported, approach. I would say a flat list of all components in the final container image. This is the approach that Anchore Syft has taken too.

But as a builder, having the dependency graph information as well is very helpful. It helps you track down where exactly that vulnerable component was pulled in from. And if a tool doesn't support dependency graph information that doesn't matter, it's a separate element to the components and is ignored.

Author: This is what the lifecycle would be currently responsible for. After the RFC at https://github.com/buildpacks/rfcs/pull/151

the idea was for the lifecycle to automatically combine the BOM while injecting metadata indicating which layer it came from.  This was meant to solve issues like https://github.com/paketo-buildpacks/packit/issues/54 automatically.

IMHO it makes more sense for a container to have a single SBOM which can have layers as its components or can be a flat list. I would assume most container scanning tools currently produce/consume a single SBOM document and dividing it up into layers might make integrations harder.  Keeping it into a single document while still preserving metadata around which layer contributed to the bom entry might be the best of both worlds imho.

Reviewer 2: From my understanding of buildpacks, each layer represents the output of each buildpack. The reason why I advocate for layer level SBOMs is because it helps with SBOM reuse on the tool side. However, the final output of kpack can combine all the SBOMs into a single one.

# Meta
[meta]: #meta
- Name: Structured SBOMs
- Start Date: 2021-06-02
- Author(s): [@samj1912](https://github.com/samj1912), [@sophiewigmore](https://github.com/sophiewigmore), [@ForestEckhardt](https://github.com/ForestEckhardt)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)

> - Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

Reviewer 5: ```suggestion
- Supersedes: [RFC 0087](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md), [RFC 0053](https://github.com/buildpacks/rfcs/blob/main/text/0053-decouple-buildpack-plan-and-bom.md)
```

# Summary
[summary]: #summary

This RFC proposes the following - 

- Use of CycloneDX as the SBOM storage format.

> - Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to a separate file.

Reviewer 5: ```suggestion
- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to separate files.
```

> - Reducing duplication of effort by standardizing formats across multiple sectors
> - Identifying suspicious or counterfeit software components

Currently since container images built by Cloud Native Buildpacks don't follow a generic file system layout, they may not be easily scannable by container scanning tools. We do however provide a BOM which should greatly help sidestep the entire manual scanning process and speed up things like CVE detection by directly providing the CVE scanner with a BOM.

However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.

Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.

## What use cases does it support?

- Consistent SBOM across buildpacks
- Improved integration with other supply chain analysis and scanning tools

## What is the expected outcome?

- Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.

> - Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to a separate file.

Reviewer 5: ```suggestion
- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to separate files.
```

# What it is
[what-it-is]: #what-it-is

Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.
It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` will be `cdx.json`

The above bom documents need to be created in CycloneDX format.

The lifecycle of the respective `bom` files during rebuilds would be exactly the same as the `bom` table currently. 

> The lifecycle would be responsible for taking all of the above `bom` files, merging it while tagging it with the originating layer or buildpack the `bom` came from and storing it in `/layers/config/bom.cdx.json`.

Reviewer 6: I am interested in the fact that all of the BOM files are merged into seemingly one BOM file. How do I distinguish things that were in the `build` BOM from those in the `launch` BOM? Is there a distinction at all in the final merged BOM? If there is no distinction why bother separating them above?

Author: There is a distinction. Each bom component entry has the following metadata properties added to it. https://github.com/samj1912/rfcs/blob/sbom/text/0000-sbom.md#how-it-works

Author: > io.buildpacks.bom.buildpack.id - Buildpack ID for the buildpack that created the BOM
> io.buildpacks.bom.types - Set to a comma seperated list of build or launch as applicable.
> io.buildpacks.bom.layer.name (Optional) - Set to the name of the layer if the bom was associated with a specific layer.

Author: Given the recent conversation in the OH - I guess we want these two to be separate files instead of just being able to distinguish between the two. Will make that change.

Reviewer 5: How we will handle SBOM merging when a build contains both older and newer buildpacks that are producing freeform toml and cyclone dx BOMs respectively?

I can't think of many elegant ways to do this so we may need to do something dramatic? e.g. maybe one of
- support two different "merged" boms if we have two types of data
- have a hard cutoff in the platform API. If platform api is 0.x or above we produce a cycloneDX bom with whatever data we have in that format. If it is below we produce a CNB BOM with the same limitations. If we do this we may want to consider allowing newer buildpacks to produce data in both formats for a time to retain functionality on older platforms.

I kinda hate both of those options. Does anyone have a better idea? If not it may be a necessary evil.

Reviewer 4: @ekcasey That's a good point I hadn't considered. I don't HATE option 2 that you've proposed, but obviously I'd love to find a state in which we don't have to support both formats. 

I think the idea of supporting two different merged BOMs in bullet # 1 would be super confusing for someone to try to look at. It would also just be kind of hard to use scanning tools if the BOMs came in two formats but were both needed to make up the entire set of metadata.

Reviewer 7: We could convert the existing free-form TOML to CycloneDX by placing the arbitrary TOML in the `properties` field of a `component`.

## Delegate SBOM merging to platform

The main reason that the lifecycle needs to have a hard dependency on CycloneDX in this proposal is because we propose that the lifecycle merges the SBOM from various buildpacks and layers. If the lifecycle were to not do this, the content of the SBOM can be completely opaque to the lifecycle. This might provide greater flexibility and almost no dependency on the lifecycle side. Instead a platform like `pack` can be responsible for gathering the individual SBOM files and merging them together for the user. This can also be provided as a reusable library/CLI so that other platforms could do the same.

# Prior Art
[prior-art]: #prior-art

- [#RFC 53](https://github.com/buildpacks/rfcs/blob/main/text/0053-decouple-buildpack-plan-and-bom.md)
- [#RFC 87](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- Implementation details about how the lifecycle should be merging the SBOMs.

> - Moving the SBOM from the `io.buildpacks.build.metadata` label to a file in the output image to avoid large labels

Reviewer 8: Is there a way to ensure that the output file (with the BOM) and the image are always linked together? Would we put the SHA of the BOM file in a label on the image?

Author: The current RFC just puts the combined BOM in a label as usual. I was hoping to have a separate RFC that moves it to a file which would do something similar (i.e. put the SHA of the BOM file in a label on the image)

However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.

Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.

## What use cases does it support?

- Consistent SBOM across buildpacks
- Improved integration with other supply chain analysis and scanning tools

## What is the expected outcome?

- Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.
- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to a separate file.
- Enhancements to the platform spec and `pack` to allow easy interaction with SBOMs including conversion between major SBOM formats.

# What it is
[what-it-is]: #what-it-is

Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.

> It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` will be `cdx.json`

Reviewer 4: What is the reason for even creating the separate files for `layer`, `launch`, and `build` if we are just going to merge them together into one file down the line? 
Is there any intent here to have separate files so that users can also access the separate BOMs?

Reviewer 7: One benefit of keeping them separate: the lifecycle could automatically insert metadata into each CycloneDX SBoM entry that maps the component to a layer digest and buildpack (and/or marks the component as build-only).

Author: Apart from what @sclevine said, each layer in a buildpack could have a different lifecycle (some could be launch only etc.). Keeping BOMs separate is an artifact of tying them to a similar lifecycle as their layers so that the BOMs and the contents they are describing don't drift apart. As for separate BOMs for diffirent buildpacks - It would get difficult for BP authors to write to a shared BOM file and random buildpacks could potentially modify BOMs for other buildpacks.

Reviewer 6: I think that your lifecycle description makes sense for the `build` and `launch` BOM files. I am still confused by the presence of the `layer` BOM file. What purpose does it serve and how would it be utilized?

Reviewer 6: What goes in the `layer` BOM that does not go into `build` or `launch` BOM? I understand the separation in terms of the "lifecycles" but I am unclear where just the `layer` BOM comes in.

Author: Some details and BG are in this RFC - https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md

The idea was for the build and launch BOM to include changes/contribution to the app/workspace and the layer bom to have the layer specific bom.

Author: > The buildpack SHOULD NOT add bom to a layer content metadata file describing the contents that are not contributed by the same layer. BOM entries that describe contribution to the application directory should go in the buildpack-level BOM entries in launch.toml or build.toml.

## What is the expected outcome?

- Use of CycloneDX and SPDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.
- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to separate files.

## Why support multiple SBOM formats?

The motivation for supporting multiple SBOM formats is because there is currently no clear leader in the SBOM space. SPDX and CycloneDX both specialize in different use cases - compliance and supply chain security respectively, and although both do have a lot of overlap, there is still loss of information when converting between the two. The other motivation is that the SBOM story doesn't end with their generation. Different users could have different SBOM consumers who support one format or the other. Leaving individual SBOM files intact allows users to post-process it later if needed without any loss in information. This allows users to flexibly use the Buildpacks API to generate a SBOM in the format that is most useful to them.

# What it is
[what-it-is]: #what-it-is

Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.

> It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` will be `cdx.json` for CycloneDX documents and `spdx.json` for SPDX documents. A buildpack can also output the `bom` in multiple formats i.e. both CycloneDX and SPDX. 

Reviewer 8: /queue-issue buildpacks/lifecycle "Builder should warn if newer buildpacks write a bom in *.toml" type/enhancement epic/sbom

Additionally the buildpack must also indicate the SBOM format(s) it exports using the in its `buildapck.toml`'s `sbom` key in the `buildpacks` table.

The format of this key should look like - 

```toml
api = "0.x"

[buildpack]
id = "<buildpack ID>"
name = "<buildpack name>"
# This can be an array of supported SBOM formats by the buildpack.
# Valid array values are "CycloneDX" and "SPDX" as of this RFC.
sbom = ["CycloneDX"]
```

This information can be displayed on the registry and `pack inspect-buildpack`.

> The `lifecycle` binaries would be responsible for taking all of the above `bom` files, and copying them in `/layers/config/sbom/<type>/<buildpack-id>/<layer-id>/bom.<ext>` for the layer `bom` files where `type` could be `build` or `launch`. Buildpack level BOM files will reside at `/layers/config/sbom/<type>/<buildpack-id>/bom.<ext>` where the type could be `launch` and `build` respectively.

Reviewer 8: /queue-issue buildpacks/lifecycle "Builder should copy bom files to /layers/config/sbom" type/enhancement epic/sbom

> Additionally the `lifecycle` binaries will be responsible for creating merged `bom` files, merging `bom` formats of the same type and outputting them at `/layers/config/<type>/sbom/bom.<ext>`. Initially this will only be supported for CycloneDX as it has a well defined and efficient way of merging multiple `bom` files. A reference implementation can be found at the [`cyclonedx-cli`](https://github.com/CycloneDX/cyclonedx-cli) project. In the future we may add support for merging SPDX `bom` files as well. 

Reviewer 8: /queue-issue buildpacks/lifecycle "Lifecycle should merge CycloneDX bom files" type/enhancement epic/sbom

Merging different formats (CycloneDX/SPDX/Legacy CNB format) is a non-goal and will not be supported. Since we will be making the individual `bom` files available as well, if users/platforms wish, they can choose to write conversion and merging logic but this will not be supported in the lifecycle.

Since we will be preserving the original output files and if a buildpack chooses to output in both SPDX and CycloneDX, the lifecycle may eventually add support for SPDX merging in a future version if merging SPDX documents becomes better defined/efficient.

# How it Works
[how-it-works]: #how-it-works

> The `lifecycle` binaries would be responsible for reading, merging and restoring the appropriate `bom` files (the legacy CNB format and the CycloneDX format currently). The lifecycle of these `bom` files would be tied to their respective metadata `toml` files. See [RFC 0087](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md#how-it-works) for more details. 

Reviewer 8: /queue-issue buildpacks/lifecycle "Exporter should cache bom files for cached layers" type/enhancement epic/sbom

Reviewer 8: /queue-issue buildpacks/lifecycle "Exporter should export bom files for launch layers" type/enhancement epic/sbom

For merging the CycloneDX `bom` files, the `lifecycle` binaries could replicate or use tooling from [CycloneDX-cli](https://github.com/CycloneDX/cyclonedx-cli) which has a merge operation. The only additional piece of information that the `lifecycle` binaries would inject are `CycloneDX` [`metadata`](https://cyclonedx.org/use-cases/#properties--name-value-store) the following property keys -

- `io.buildpacks.bom.buildpack.id` - Buildpack ID for the buildpack that created the BOM

> - `io.buildpacks.bom.layer.name` (Optional) - Set to the name of the layer if the `bom` was associated with a specific layer. 

Reviewer 8: /queue-issue buildpacks/lifecycle "Exporter should add io.buildpacks.bom.* labels" type/enhancement epic/sbom

The `lifecycle` binaries will put the `bom` files from layers set to `launch = true` inside the `/layers/config/sbom/launch/<buildpack-id>/<layer-id>/bom.<ext>` hierarchy otherwise it will put them in `/layers/config/sbom/build/<buildpack-id>/<layer-id>/bom.<ext>`.

> ## BOM Restoration

Reviewer 8: /queue-issue buildpacks/lifecycle "Restorer should restore bom files from app and cache" type/enhancement epic/sbom

# Backwards compatibility

Since we will not be merging the new formats with the legacy CNB format, we can just output the merged CNB formatted SBOM as a separate file.

# Drawbacks
[drawbacks]: #drawbacks

This RFC adds a fair bit of complexity to the lifecycle and makes it responsible for more than just buildpack orchestration and container/layer assembly.

# Alternatives
[alternatives]: #alternatives

Only use one SBOM format like `CycloneDX` or `SPDX` and bake it into the lifecycle.

# Prior Art
[prior-art]: #prior-art

- [#RFC 53](https://github.com/buildpacks/rfcs/blob/main/text/0053-decouple-buildpack-plan-and-bom.md)
- [#RFC 87](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- Interactions with the stack SBOM and how to represent that and merge it with the Buildpack SBOM.

> - Restore/export logic for SBOM files. This will be covered in a subsequent RFC.

Reviewer 9: With the recent changes from talking with @ekcasey, is this still true @samj1912?

Author: Nope - should remove that.

[what-it-is]: #what-it-is

Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.
It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` will be `cdx.json` for CycloneDX documents and `spdx.json` for SPDX documents. A buildpack can also output the `bom` in multiple formats i.e. both CycloneDX and SPDX. 

Additionally the buildpack must also indicate the SBOM format(s) it exports using the in its `buildapck.toml`'s `sbom` key in the `buildpacks` table.

The format of this key should look like - 

```toml
api = "0.x"

[buildpack]
id = "<buildpack ID>"
name = "<buildpack name>"
# This can be an array of supported SBOM formats by the buildpack.
# Valid array values are "CycloneDX" and "SPDX" as of this RFC.
sbom = ["CycloneDX"]
```

This information can be displayed on the registry and `pack inspect-buildpack`.

The lifecycle of the respective `bom` files during rebuilds would be exactly the same as the `bom` table currently. 

> The lifecycle would be responsible for taking all of the above `bom` files, and putting them in `/layers/config/sbom/<buildpack-id>/<layer-id>/<type>.bom.<ext>` for the layer `bom` files where `type` call be `build` or `launch` or `both`. Buildpack level BOM files will reside at `/layers/config/sbom/<buildpack-id>/build.bom.<ext>` and `/layers/config/sbom/<buildpack-id>/launch.bom.<ext>` for the `launch` and `build` `bom` respectively.

Reviewer 8: Just to call it out explicitly, `type` would be determined by what is the value under the `[types]` table in the `<layer>.toml` file?

Reviewer 8: Are we still putting bom information in `io.buildpacks.build.metadata`? If not, would it make sense for each `/layers/config/sbom/<buildpack-id>/<layer-id>/<type>.bom.<ext>` to be its own layer, to aid in layer re-use?

Could we translate this "[how it works](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md#how-it-works)" from the bom-in-layer-metadata RFC into this new world?

A couple of thoughts:
* If the layer is `cache = true` and there is a `<layer>.bom.<ext>` then the bom would be saved in the cache alongside the layer and restored to the buildpack's `<layers>` directory on the next build
* If the layer is `launch = true`, `cache = false`, `build = false` and there is a `<layer>.bom.<ext>` then normally we are only restoring the layer metadata, but would the buildpack also want the bom to be available, so that it could decide whether to re-use the layer or not? In this case we would need to copy the `/layers/config/sbom/<buildpack-id>/<layer-id>/<type>.bom.<ext>` from the previous image as well (unless this is going in a label?

Reviewer 6: I feel like in the case of `launch = true`, `cache = false`, `build = false` I would want the `<layer>.bom.<ext>` or at least the information therein to be available so that I have the option to reuse the BOM and not generate a new one in the case that I am reusing that layer.

That is my two cents on the matter at least.

Author: This is the part that @ekcasey and I need to sync up about.

# How it Works
[how-it-works]: #how-it-works

The lifecycle would be responsible for reading, merging and restoring the appropriate `bom` files (the legacy CNB format and the CycloneDX format currently). The lifecycle of these `bom` files would be tied to their respective metadata `toml` files. See [RFC 0087](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md#how-it-works) for more details. 

For merging the CycloneDX `bom` files, the lifecycle could replicate or use tooling from [CycloneDX-cli](https://github.com/CycloneDX/cyclonedx-cli) which has a merge operation. The only additional piece of information that the lifecycle would inject are `CycloneDX` [`metadata`](https://cyclonedx.org/use-cases/#properties--name-value-store) the following property keys -

- `io.buildpacks.bom.buildpack.id` - Buildpack ID for the buildpack that created the BOM
- `io.buildpacks.bom.layer.name` (Optional) - Set to the name of the layer if the `bom` was associated with a specific layer. 

> The lifecycle will put all the `build` related entries in the `/layers/config/sbom/build.bom.cdx.json` file and `launch` entries in `/layers/config/sbom/launch.bom.cdx.json`

Reviewer 8: > Do these end up in `/layers/config/sbom/launch.bom.cdx.json`?

If the answer to this question is yes, then I would think that any layer that has a `<layer>.bom.<ext>` file but is `launch = false` would have the bom end up in `/layers/config/sbom/build.bom.cdx.json`. Is that right?

> The lifecycle will put all the `build` related entries in the `/layers/config/sbom/build.bom.cdx.json` file and `launch` entries in `/layers/config/sbom/launch.bom.cdx.json`

Author: Yup that is correct as of now. I wonder if we should allow buildpack authors more control over this and let them create the the appropriate files with the extension for clarity + control?

> The lifecycle will put all the `build` related entries in the `/layers/config/sbom/build.bom.cdx.json` file and `launch` entries in `/layers/config/sbom/launch.bom.cdx.json`

Author: @natalieparellano clarified this a bit more. LMK if it makes sense.

# Meta
[meta]: #meta
- Name: Structured SBOMs
- Start Date: 2021-06-02
- Author(s): [@samj1912](https://github.com/samj1912), [@sophiewigmore](https://github.com/sophiewigmore), [@ForestEckhardt](https://github.com/ForestEckhardt)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

This RFC proposes the following - 

- Use of CycloneDX as the SBOM storage format.
- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to separate files.

# Definitions
[definitions]: #definitions

> - SBOM/BOM: A software bill of materials (SBOM) is a list of components in a piece of software. Software vendors often create products by assembling open source and commercial software components. The SBOM describes the components in a product. In case of buildpacks the SBOM describes the contents of the various layers, buildpacks, stacks and the output app container.

Reviewer 10: ```suggestion
- SBOM (Software Bill Of Materials) / BOM: A list of components in a piece of software. Software vendors often create products by assembling open source and commercial software components. The SBOM describes the components in a product. In case of buildpacks the SBOM describes the contents of the various layers, buildpacks, stacks and the output app container.
```

> - CycloneDX: CycloneDX is a lightweight software bill of materials (SBOM) standard designed for use in application security contexts and supply chain component analysis.

Reviewer 10: ```suggestion
- CycloneDX: A lightweight SBOM standard designed for use in application security contexts and supply chain component analysis.
```

> - SPDX: Software Package Data Exchange (SPDX) is a file format used to document information on the software licenses under which a given piece of computer software is distributed.

Reviewer 10: ```suggestion
- SPDX (Software Package Data Exchange): A file format used to document information on the software licenses under which a given piece of computer software is distributed.
```

> - SWID: SWID is a formal industry standard used by various commercial software publishers designed with software inventory and entitlements management in mind.

Reviewer 10: ```suggestion
- SWID: A formal industry standard used by various commercial software publishers designed with software inventory and entitlements management in mind.
```

# How it Works
[how-it-works]: #how-it-works

The lifecycle would be responsible for reading, merging and restoring the appropriate `bom` files. The lifecycle of these `bom` files would be tied to their respectice metadata `toml` files. See [RFC 0087](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md#how-it-works) for more details. For merging the `bom` files, the lifecycle could replicate or use tooling from [CycloneDX-cli](https://github.com/CycloneDX/cyclonedx-cli) which has a merge operation. The only additional piece of information that the lifecycle would inject are `CycloneDX` [`metadata`](https://cyclonedx.org/use-cases/#properties--name-value-store) the following property keys -

- `io.buildpacks.bom.buildpack.id` - Buildpack ID for the buildpack that created the BOM
- `io.buildpacks.bom.layer.name` (Optional) - Set to the name of the layer if the `bom` was associated with a specific layer. 

> Thw lifecycle will put all the `build` related entries in the `/layers/config/build.bom.cdx.json` file and `launch` entries in `/layers/config/launch.bom.cdx.json`

Reviewer 10: ```suggestion
The lifecycle will put all the `build` related entries in the `/layers/config/build.bom.cdx.json` file and `launch` entries in `/layers/config/launch.bom.cdx.json`
```

# Backwards compatibility

For `BOM` from older buildpacks the following mappings will be defined. Each `BOM` entry will be added as a `CycloneDX` component in the final components list.

For eg. if we have the BOM table like below -

```toml
[[bom]]
name = "example"
[bom.metadata]
version = "0.2.3"
foo = "bar"
foo_array = ["bar"]
[[bom]]
name = "example2"
[bom.metadata]
version = "0.0.1"
foo = "bar"
```

> It will be converted to - 

Reviewer 10: Maybe it's worth adding when it's going to be converted and who is responsible to do so.
